pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("cannot retrieve cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+	}
+
+	if cluster == nil {
+		// cluster does not exist, it needs to be created
+		cluster, err = buildCluster(clusterDeployment)
+		if err != nil {
+			return fmt.Errorf("cannot build cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		_, err := c.capiClient.ClusterV1alpha1().Clusters(clusterDeployment.Namespace).Create(cluster)
+		if errors.IsConflict(err) {
+			// cluster was created elsewhere, no need to retry
+			clustoplog.WithClusterDeployment(c.logger, clusterDeployment).Debugf("conflict creating cluster, skipping")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error creating cluster for cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		return nil
+	}
+
+	// cluster exists, make sure it reflects the current cluster deployment spec
+	deploymentConfigSpec, err := controller.ClusterDeploymentSpecFromCluster(cluster)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve a cluster deployment spec from existing cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+	}
+	if !apiequality.Semantic.DeepEqual(deploymentConfigSpec, clusterDeployment.Spec) {
+		providerConfig, err := controller.ClusterProviderConfigSpecFromClusterDeploymentSpec(&clusterDeployment.Spec)
+		if err != nil {
+			return fmt.Errorf("cannot create a cluster provider config spec from cluster deployment %s/%s: %v", clusterDeployment.Namespace, clusterDeployment.Name, err)
+		}
+		updatedCluster := cluster.DeepCopy()
+		updatedCluster.Spec.ProviderConfig.Value = providerConfig
+		_, err = c.capiClient.ClusterV1alpha1().Clusters(updatedCluster.Namespace).Update(updatedCluster)
+		if err != nil {
+			return fmt.Errorf("cannot update existing cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+		}
+	}
+	return nil
+}
+
+// syncControlPlane takes a cluster deployment and ensures that a corresponding master machine set
+// exists and that its spec reflects the spec of the master machineset in the cluster deployment spec.
+func (c *Controller) syncControlPlane(clusterDeployment *clustop.ClusterDeployment, clusterVersion *clustop.ClusterVersion) error {
+	machineSetName := masterMachineSetName(clusterDeployment.Name)
+	machineSet, err := c.machineSetsLister.MachineSets(clusterDeployment.Namespace).Get(machineSetName)
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("error retrieving master machineset %s/%s: %v", clusterDeployment.Namespace, machineSetName, err)
+	}
+
+	if machineSet == nil {
+		// machine set does not exist, it needs to be created
+		machineSet, err = buildMasterMachineSet(clusterDeployment, clusterVersion)
+		c.logger.Debugf("Just built machineset: %#v", machineSet)
```

Comment:
I'd be tempted to create or pass around a contextual logger to each of these methods.